feat: Initial zkSolc support

Karrq opened the pull request:
Add initial zksolc support to compilers:
* use zksolc instead of solc
* parse and write zksolc-compatible compiler input & output

Karrq committed:
refactor: Solc -> ZkSolc

feat(zksolc): use zksolc binary

Karrq committed:
fix: zksolc version parser

Karrq committed:
feat: zksolc compile output (wip)

Karrq committed:
fix: lints

Karrq committed:
feat: zk or solc metadata in contracts

Karrq committed:
fix: bytecodehash keccak256

Karrq committed:
fix: optional versions CompilerOutput

Karrq committed:
tests: fix doctests compilation

docs: change type of 1 Remappings codeblock to be solidity and not an
ignored test

Karrq committed:
Merge remote-tracking branch 'origin/main' into feat/zksolc

hedgar2017 commented on src/artifacts/contract.rs:
```diff
@@ -4,42 +4,56 @@ use crate::artifacts::{
+
+    pub metadata: Option<ZkOrSolcMetadata>,
+
+    //TODO: remove?
```
These you can request. Here is [the list](https://github.com/matter-labs/era-compiler-solidity/blob/main/src/solc/standard_json/input/settings/selection/file/flag.rs) of output fields we support.

hedgar2017 commented on src/artifacts/contract.rs:
```diff
@@ -4,42 +4,56 @@ use crate::artifacts::{
     #[serde(default, skip_serializing_if = "Option::is_none")]
     pub evm: Option<Evm>,
+
+    //TODO: remove?
```
This we don't support though.

hedgar2017 commented on src/artifacts/mod.rs:
```diff
@@ -1049,6 +1049,7 @@ pub enum BytecodeHash {
     Ipfs,
     None,
     Bzzr1,
+    Keccak256,
```
We do not support `Ipfs` and `Bzzr1` either, just in case this enum is for `zksolc` only.

Karrq committed:
wip

feat: Zk CompilerInput
feat: ZkSettings, ZkStandardJsonInput, OutputSelection & Sources conversion

Karrq committed:
fix: test compilation

Karrq committed:
wip outputs

feat: zk compiler output conversion
wip(refacto)r: contract -> zkContract

Karrq committed:
wip more outputs (compile success)

elfedy commented:
Closing in favor of https://github.com/Moonsong-Labs/compilers/pull/2
